### PR TITLE
Vite: Add partial SvelteKit support

### PR DESCRIPTION
**Pull request by JReinhold**
Issue: #19280

This PR is a first stab at setting up a SvelteKit repro and getting it to work out of the box.

## TODO

- [x] Test that base `vite-svelte` framework still works
- [x] Test that TS SvelteKit framework works.
- [x] Test that it works when published to verdaccio

## What I did

- Add repros via [`create-svelte-with-args`](https://github.com/storybookjs/create-svelte-with-args)
- Removed the `sveltePreprocess` option from the template as that was causing ESM vs CJS issues and we decided it was an anti-pattern anyway since they should automatically be picked up by the Vite config
- Remove `vite-plugin-svelte-kit` whenever it is recognised in a Vite config because it conflicts with Storybook's VIte server handling. This has some drawbacks that certain SvelteKit features aren't available in Storybook but for now we consider that a general issue that needs to be solved in SvelteKit.

## How to test

1. `git checkout origin/jeppe/sb-544-sveltekit`
2. Use Node 16 instead of Node 14 (SvelteKit can't be built using v14) - eg. `nvm use 16`
3. Generate repro with `cd code && yarn generate-repros-next --template svelte-kit/skeleton-js`
4. `cd ../sandbox/svelte-kit-skeleton-js && yarn install`
5. `yarn storybook`
6. See that all basic stories work.
7. Repeat for SvelteKit with TypeScript: `svelte-kit/skeleton-ts`

## Out of scope

I want to include a few stories based on the Svelte CSF format in the sandbox so CI also ensures that they work. That's another PR.

### CODE REVIEW

File: code/frameworks/svelte-vite/src/preset.ts
Comment (JReinhold) [inline code/frameworks/svelte-vite/src/preset.ts:28]: Ideally this should have been a plugin, but it doesn't seem to have an affect. The plugin correctly removes `vite-plugin-svelte-kit` from the config but it doesn't  change any behavior

File: code/lib/builder-vite/src/vite-config.ts
Comment (JReinhold) [inline code/lib/builder-vite/src/vite-config.ts:100]: modified a bit from the original implementation that only allowed `.storybook`, since doing that would give errors about files in `template` not being allowed. That might be a detail only relevant when sandboxing, but still something we want to work.

This is strictly not necessary for supporting SvelteKit as removing the whole `vite-plugin-svelte-kit` makes this unnecessary.

File: code/lib/builder-vite/src/vite-server.ts
Comment (JReinhold) [inline code/lib/builder-vite/src/vite-server.ts:27]: only minor refactor, same logic.

File: code/lib/cli/src/generators/SVELTE/index.ts
Comment (JReinhold) [inline code/lib/cli/src/generators/SVELTE/index.ts:34]: as per https://github.com/storybookjs/storybook/issues/19280#issuecomment-1264132087

File: code/lib/builder-vite/src/vite-config.ts
Comment (benmccann) [reply code/lib/builder-vite/src/vite-config.ts:100]: Could we just include `.storybook` and `template`? We don't necessarily want to include `.` as some users were worried about exposing `.env` files, etc.

File: scripts/sandbox.ts
Comment (benmccann) [inline scripts/sandbox.ts:16]: this doesn't appear to be used

File: code/lib/builder-vite/src/vite-config.ts
Comment (JReinhold) [reply code/lib/builder-vite/src/vite-config.ts]: That makes sense. Done.

File: code/frameworks/svelte-vite/src/preset.ts
Comment (IanVS) [inline code/frameworks/svelte-vite/src/preset.ts:24]: I'd still love to find a way to avoid having to do this, since we'll start getting complaints about sveltekit aliases not working (even though there seems to be some debate over whether they should be used at all inside components).  But, this seems like a reasonable workaround for now, until we can figure it out.

File: code/lib/builder-vite/src/vite-server.ts
Comment (IanVS) [reply code/lib/builder-vite/src/vite-server.ts:27]: This change means that the `appType` and `server` options are not passed to `getOptimizedDeps()`.  I think this is okay, since we're just passing the config in so that we can create a resolver against which to check the candidates for pre-optimization.

**Review by IanVS — COMMENTED** [review]
Thanks for this!  It'll be great to have sveltekit repros, and getting it working out-of-the-box is an important step to better svelte support!

File: code/lib/builder-vite/src/vite-server.ts
Comment (JReinhold) [reply code/lib/builder-vite/src/vite-server.ts:27]: good catch, hadn't thought about that.

File: code/frameworks/svelte-vite/src/preset.ts
Comment (JReinhold) [reply code/frameworks/svelte-vite/src/preset.ts:24]: I agree, let's continue investigating the core issue and do a follow up to this.

File: code/lib/builder-vite/src/vite-config.ts
Comment (JReinhold) [reply code/lib/builder-vite/src/vite-config.ts:94]: `template-stories` gets copied over when creating a new sandbox, it contains generic stories for testing addons and stuff. So it's only relevant when sandboxing, not in a real use-case.

Comment (IanVS) [reply code/lib/builder-vite/src/vite-config.ts:94]: I think then maybe it belongs in the `viteFinal` of the sandbox projects, instead of here, which would be added to every user project.

Comment (JReinhold) [reply code/lib/builder-vite/src/vite-config.ts]: You're right, I hadn't even considered that an option.

File: code/lib/builder-vite/src/vite-config.ts
Comment (JReinhold) [reply code/lib/builder-vite/src/vite-config.ts:86]: oh yeah that was a typo, you're absolutely right.

File: scripts/sandbox.ts
Comment (JReinhold) [reply scripts/sandbox.ts:269]: yeah I tried that but you actually can't because that won't pick up on other plugin's configs.
So if I set it directly on the config `config?.server?.fs?.allow` will be falsy, while doing it through this plugin logic `config?.server?.fs?.allow` will actually include `vite-plugin-svelte-kit`'s allow list.

Comment (IanVS) [reply scripts/sandbox.ts:269]: I meant without the check, you could just set the allow value directly, and then the rest of everything should still work.  `vite-plugin-svelte-kit` should extend the user's config and the other plugin you created should as well, right?  If we can't set custom allow paths in `viteFinal`, we're in trouble.

Comment (benmccann) [reply scripts/sandbox.ts:269]: I think the plugin is probably the safer approach because it can check whether the user has set `server.fs.allow` themselves or not. If we always pass `server.fs.allow: ['.storybook']` then we enable that feature which will may break the user's project if they have not added the directories their project uses to the config.

Comment (IanVS) [reply scripts/sandbox.ts:269]: In this sandbox, we _are_ the user.  I'm not talking about setting it in all cases, just in the example sandbox, which is used for testing.

Comment (benmccann) [reply scripts/sandbox.ts:269]: oh yeah, that sounds like the better approach then

File: code/frameworks/svelte-vite/src/preset.ts
Comment (JReinhold) [reply code/frameworks/svelte-vite/src/preset.ts:35]: oh yeah, I hadn't even noticed that.
I have a version that is more declarative, but I'm unsure if it's actually more readable. But I don't have any strong opinions so if you think it's better I'll happily push it.

```ts
export const viteFinal: StorybookConfig['viteFinal'] = async (config, { presets }) => {
  const { plugins = [] } = config;

  return {
    ...config,
    plugins: [
      // remove SvelteKit plugin
      ...plugins.map(withoutSvelteKitPlugin),
      // Add svelte plugin if not present
      !hasPlugin(plugins, 'vite-plugin-svelte') &&
        (await import('@sveltejs/vite-plugin-svelte')).svelte(),
      // Add docgen plugin
      svelteDocgen(config),
    ].filter(Boolean),
  };
};

const withoutSvelteKitPlugin = (plugin: PluginOption): PluginOption => {
  if (Array.isArray(plugin)) {
    // recursive - Vite plugins can be nested
    return plugin.map(withoutSvelteKitPlugin).filter(Boolean);
  }
  if (plugin && 'name' in plugin && plugin.name === 'vite-plugin-svelte-kit') {
    return undefined;
  }
  return plugin;
};
```

Comment (IanVS) [reply code/frameworks/svelte-vite/src/preset.ts:35]: Thanks, the way you've got it right now is fine I think, and hopefully this is temporary anyway, until we can figure out what's causing the app to load instead of storybook in 7.0.

**Review by IanVS — COMMENTED** [review]
This looks good to me, but I see CI failures like:

>Unable to find the Storybook folder in "/tmp/storybook/sandbox/svelte-kit-skeleton-js/.storybook". Are you sure it exists? Or maybe this folder uses a custom Storybook config directory?

Is that because the new repros have not yet been published?

**Comment by JReinhold** [comment]
> Is that because the new repros have not yet been published?

Yes I think that's probably it, they should work once we merge, or alternatively I can manually publish them - but I'll hold off until Node 16+ is ready in the repo.

File: code/frameworks/angular/src/server/framework-preset-angular-ivy.ts
Comment (IanVS) [inline code/frameworks/angular/src/server/framework-preset-angular-ivy.ts:61]: This seems unrelated to this PR, but I'm guessing it'll fall back out when `next` is updated and merged in.

**Review by IanVS — APPROVED** [review]

**Comment by benmccann** [comment]
@JReinhold just FYI, it looks like there's a merge conflict

File: code/lib/cli/src/generators/SVELTE/index.ts
Comment (shilman) [reply code/lib/cli/src/generators/SVELTE/index.ts:34]: Does this need any MIGRATION.md notes? @JReinhold

Comment (JReinhold) [reply code/lib/cli/src/generators/SVELTE/index.ts:34]: You're right, users needs to remove their `svelteOptions` property in `main.cjs` and rely on their Vite config to pick them up instead. Svelte+Webpack users probably don't want to remove anything.

I'll write it up.

Comment (IanVS) [reply code/lib/cli/src/generators/SVELTE/index.ts:34]: What do you think about pulling this change out to its own PR?  I'd love to get it into the next release, since svelte apps are currently broken after bootstrapping with Storybook.

**Comment by IanVS** [comment]
This looks good, but as far as I know, `build-storybook` will still be broken when using svelteKit (see https://github.com/storybookjs/storybook/issues/19280#issuecomment-1277810919).

Maybe we can merge this in, but not advertise sveltekit too hard until we can take care of that part.

**Comment by JReinhold** [comment]
Right, that looks gnarly.
It really seems like were struggling to incorporate a metaframework that (rightfully) wants to be it's own server and builder and isn't really built for others to hook into that (because why would it).

This will probably not be the last time facing these issues (hello SolidStart, Astro, Qwik City and Friends), so I wonder if we could think about this in a broader way..

But let's get this in and fix building the Storybook later.

**Comment by IanVS** [comment]
Awesome, thanks for slogging through and getting this in, @JReinhold!

**Comment by shilman** [comment]
@JReinhold FWIW I would call this a `maintenance` PR because it touches user-facing code in the Vite builder, and in a significant way.

**Pull request merged**
